rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,999 @@
+    }
+
+    /** Deserialization [[ByteVector]] to [[Node]]
+      */
+    def decode(bv: ByteVector): Node = {
+      val arr     = bv.toArray
+      val maxSize = arr.length
+
+      // If first bit 0 - return true, otherwise false.
+      def isLeaf(secondByte: Byte) = (secondByte & 0x80) == 0x00
+
+      @tailrec
+      // Each loop decodes one non-empty item.
+      def decodeItem(pos0: Int, node: Node): Node =
+        if (pos0 == maxSize) node // End of deserialization
+        else {
+          val (idx0Next, nodeNext) = try {
+            val numItem: Int = byteToInt(arr(pos0)) // Take first byte - it's item's number
+            assert(
+              node(numItem) == EmptyItem,
+              "Error during deserialization: wrong number of item."
+            )
+            val pos1       = pos0 + 1
+            val secondByte = arr(pos1) // Take second byte
+
+            // Decoding prefix
+            val prefixSize: Int = secondByte & 0x7F // Lower 7 bits - it's size of prefix (0..127).
+            val prefix          = new Array[Byte](prefixSize)
+            val posPrefixStart  = pos1 + 1
+            for (i <- 0 until prefixSize) prefix(i) = arr(posPrefixStart + i) // Take prefix
+
+            // Decoding leaf or nodePtr data
+            val valOrPtr         = new Array[Byte](defSize)
+            val posValOrPtrStart = posPrefixStart + prefixSize
+            for (i <- 0 until defSize)
+              valOrPtr(i) = arr(posValOrPtrStart + i) // Take next 32 bytes - it's data
+
+            val pos0Next = posValOrPtrStart + defSize // Calculating start position for next loop
+
+            // Decoding type of non-empty item
+            val item = if (isLeaf(secondByte)) {
+              Leaf(ByteVector(prefix), ByteVector(valOrPtr))
+            } else {
+              NodePtr(ByteVector(prefix), ByteVector(valOrPtr))
+            }
+
+            (pos0Next, node.updated(numItem, item))
+          } catch {
+            case _: Exception =>
+              assert(assertion = false, "Error during deserialization: out of range")
+              (maxSize, emptyNode)
+          }
+          decodeItem(idx0Next, nodeNext) // Try to decode next item.
+        }
+
+      decodeItem(0, emptyNode)
+    }
+  }
+
+  /**
+    * Find the common part of b1 and b2.
+    *
+    * @return (Common part , rest of b1, rest of b2).
+    */
+  def commonPrefix(vL: ByteVector, vR: ByteVector): (ByteVector, ByteVector, ByteVector) = {
+    val lengthL   = vL.length
+    val lengthR   = vR.length
+    val lengthMin = lengthL min lengthR
+
+    @tailrec
+    def findLastCommonIdx(idx: Long): Long =
+      if (idx >= lengthMin) idx
+      else if (vL(idx) == vR(idx)) findLastCommonIdx(idx + 1)
+      else idx
+
+    val lastCommonIdx = findLastCommonIdx(0)
+    val common        = vL.take(lastCommonIdx)
+    val restL         = vL.drop(lastCommonIdx)
+    val restR         = vR.drop(lastCommonIdx)
+    (common, restL, restR)
```

Comment:
This code now has multiple iterations of both vectors. Each take or drop (see ByteVector implementation) will iterate the same part of the vector.
Ideally, we want to iterate only once and collect all data that is needed.

Reply:
Thanks, it's my mistake. I thought that take-and-drop - fast operations. I roll back this change. I can use an array.slice(), but don't want to use this, because we will spend time for convert operations.

Reply:
Thanks, it's my mistake. I thought that take-and-drop - fast operations. I roll back this change. I can use an array.slice, but don't want to use this, because we will spend time for convert operations.